Strip path from AsEnumerable results

HaoK opened the pull request:
Fixes https://github.com/aspnet/Configuration/issues/473

@divega

davidfowl commented:
This is a breaking change right?

cesarbs commented on test/Microsoft.Extensions.Configuration.Test/ConfigurationTest.cs:
```diff
@@ -92,14 +92,65 @@ public void AsEnumerateFlattensIntoDictionaryTest()
+            configurationBuilder.Add(memConfigSrc3);
+
+            var config = configurationBuilder.Build();
+
```
nit: add `// Assert` comment

cesarbs commented on test/Microsoft.Extensions.Configuration.Test/ConfigurationTest.cs:
```diff
@@ -92,14 +92,65 @@ public void AsEnumerateFlattensIntoDictionaryTest()
         }
 
+        [Fact]
+        public void AsEnumerateStripsKeyFromChildren()
```
Test something like `{"Mem0", "ValueInMem0"}` i.e. a key-value pair on the config root. What's the expected behavior? Won't there be an off-by-one error in `var prefixLength = rootSection != null ? rootSection.Path.Length + 1 : 0;`?

HaoK replied:
config != section guards against that, but, I think that does reveal a subtle issue where we no longer display the value of the current node as a result

HaoK replied:
Should we print out something awk like empty string : value for this situation?

divega replied:
Are you guys talking about a value at the section where the call is made? If yes, it is interesting indeed. I don't think we ever look for this at the actual configuration root, and I am not sure `AsEnumerable()` should even return it when called on a section. Does it do it now? Re empty string:value, be careful, that empty string is a valid child key. The path of the root is null, not empty string.

divega commented:
Not against having a bool parameter on `AsEnumerable()` to opt in thus behavior (I don't think the current behavior is necessarily wrong, just unexpected depending on what you are doing). Read naming, the high level semantics is that paths returned become relative. The fact that we are stripping prefixes is an implementation detail.

HaoK commented:
Updated, adding removePathFromChildKeys to opt into the new behavior, added a few more tests verifying that the value is not shown for children, and also empty keys are supported

HaoK commented:
Also added a test to confirm that this flag does nothing when called on the root (since it has no path)

HaoK commented on src/Microsoft.Extensions.Configuration.Abstractions/ConfigurationExtensions.cs:
```diff
@@ -26,17 +26,28 @@ public static string GetConnectionString(this IConfiguration configuration, stri
             var stack = new Stack<IConfiguration>();
             stack.Push(configuration);
+            var rootSection = configuration as IConfigurationSection;
+            var prefixLength = (removePathFromChildKeys && rootSection != null) ? rootSection.Path.Length + 1 : 0;
```
Note: this code is kind of ugly, if we represented the Root as simply having a null/empty path with the property on IConfiguration, we wouldn't have to always do this cast to IConfigurationSection every time we do traversal code on config looking for a Path...

HaoK commented on src/Microsoft.Extensions.Configuration.Abstractions/ConfigurationExtensions.cs:
```diff
@@ -26,17 +26,28 @@ public static string GetConnectionString(this IConfiguration configuration, stri
         /// <returns>An enumeration of key value pairs.</returns>
-        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration)
+        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration) => configuration.AsEnumerable(removePathFromChildKeys: false);
+
```
Think about a new name @divega

divega replied:
@HaoK The best names I have come up with are `includePathFromRoot` (true by default), `trimPathFromRoot` (false by default) and `useRelativePaths` (false by default). Naming is hard. I will give it some sleep :smile:

HaoK replied:
I'm split, I like include or trim as the prefix, as opposed to use which doesn't really convey intent.

But dislike PathFromRoot, and prefer either Relative/Absolute Path.

includeAbsolutePath, trimAbsolutePath, makePathsRelative??

HaoK replied:
makePathsRelative is what we decided on

HaoK commented on src/Microsoft.Extensions.Configuration.Abstractions/ConfigurationExtensions.cs:
```diff
@@ -26,17 +26,28 @@ public static string GetConnectionString(this IConfiguration configuration, stri
+        /// Get the enumeration of key value pairs within the <see cref="IConfiguration" />
+        /// </summary>
+        /// <param name="configuration">The <see cref="IConfiguration"/> to enumerate.</param>
+        /// <param name="removePathFromChildKeys">If true, removes the configuration path from the child keys.</param>
```
If true, the child keys returned will have the current configuration's Path trimmed from the front.

HaoK commented:
c202d3148130cacadb6133425e14ac134bf9cb68
